src/H5Dvirtual.c

```diff
@@ -3672,3 +3893,454 @@ H5D__virtual_release_source_dset_files(H5D_virtual_held_file_t *head)
+            if (H5D__virtual_read_one_src(dset_info, &mapping->sub_dset[j]) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "unable to read source dataset");
+    } /* end if */
+    else
+        /* Read from source dataset */
+        if (H5D__virtual_read_one_src(dset_info, &mapping->source_dset) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "unable to read source dataset");
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5D__virtual_read_one_mapping() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5D__virtual_write_one_mapping
+ *
+ * Purpose:     Write to a single mapping entry in a virtual dataset
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__virtual_write_one_mapping(H5D_dset_io_info_t *dset_info, H5O_storage_virtual_ent_t *mapping)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE
+
+    /* Sanity check that virtual space has been patched by now */
+    assert(mapping->virtual_space_status == H5O_VIRTUAL_STATUS_CORRECT);
+
+    /* Check for "printf" source dataset resolution */
+    if (mapping->psfn_nsubs || mapping->psdn_nsubs) {
+        /* Iterate over sub-source dsets */
+        for (size_t j = mapping->sub_dset_io_start; j < mapping->sub_dset_io_end; j++)
+            if (H5D__virtual_write_one_src(dset_info, &mapping->sub_dset[j]) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "unable to write to source dataset");
+    }
+    else
+        /* Write to source dataset */
+        if (H5D__virtual_write_one_src(dset_info, &mapping->source_dset) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "unable to write to source dataset");
+
+done:
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5D__virtual_write_one_mapping() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5D__virtual_close_mapping
+ *
+ * Purpose:     Frees memory structures allocated by H5D__virtual_pre_io
+ *              for a particular mapping
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__virtual_close_mapping(H5O_storage_virtual_ent_t *mapping)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE
+
+    /* Check for "printf" source dataset resolution */
+    if (mapping->psfn_nsubs || mapping->psdn_nsubs) {
+        /* Iterate over sub-source dsets */
+        for (size_t j = mapping->sub_dset_io_start; j < mapping->sub_dset_io_end; j++)
+            /* Close projected memory space */
+            if (mapping->sub_dset[j].projected_mem_space) {
+                if (H5S_close(mapping->sub_dset[j].projected_mem_space) < 0)
+                    HGOTO_ERROR(H5E_DATASET, H5E_CLOSEERROR, FAIL, "can't close temporary space");
```

Comment:
Previously, in `H5D__virtual_post_io()`, these `H5S_close()` calls used HDONE_ERROR, presumably to make sure all the dataspaces get closed even on error.